icu4c/source/common/locdspnm.cpp

```diff
@@ -723,11 +723,25 @@ LocaleDisplayNamesImpl::localeIdName(const char* localeId,
             return result;
         }
     }
-    if (substitute) {
-        return langData.get("Languages", localeId, result);
-    } else {
-        return langData.getNoFallback("Languages", localeId, result);
+    langData.getNoFallback("Languages", localeId, result);
+    if (result.isBogus() && uprv_strchr(localeId, '_') == NULL) {
```

Comment:
Because converting to canonical *form* (as opposed to canonical *syntax*) can do things like map aliased language codes, e.g. "prs" to "fa_AF". That is the whole point of this fix...